enhancements/kube-apiserver/kms-encryption-foundations.md

```diff
@@ -0,0 +1,384 @@
+
+**stateController** generates EncryptionConfiguration for API server consumption. Implements distributed state machine ensuring all API servers converge to same revision. For KMS mode, generates configuration with deterministic Unix socket paths.
+
+**migrationController** orchestrates resource re-encryption. Marks resources as migrated after rewriting in etcd. Works with all encryption modes including KMS.
+
+**pruneController** prunes inactive encryption key secrets. Maintains N keys (currently 10) for rollback scenarios.
+
+**conditionController** determines when controllers should act. Provides status conditions (`EncryptionInProgress`, `EncryptionCompleted`, `EncryptionDegraded`).
+
+#### Steps for Enabling KMS Encryption
+
+1. Cluster admin updates the APIServer resource:
+   ```yaml
+   apiVersion: config.openshift.io/v1
+   kind: APIServer
+   spec:
+     encryption:
+       type: kms
+       kms:
+         aws:
+           region: us-east-1
+           keyArn: arn:aws:kms:us-east-1:123456789012:key/12345678-1234-1234-1234-123456789012
+   ```
+
+2. keyController detects the new encryption mode and calculates hash of the KMS configuration.
+
+3. keyController creates encryption key secret:
+   ```yaml
+   apiVersion: v1
+   kind: Secret
+   metadata:
+     name: openshift-kube-apiserver-encryption-1
+     namespace: openshift-config-managed
+     annotations:
+       encryption.apiserver.operator.openshift.io/mode: "kms"
+       encryption.apiserver.operator.openshift.io/kms-config-hash: "a1b2c3d4e5f67890"
+   data:
+     keys: ""  # Empty in Tech Preview - KEK stored in external KMS
+              # In Tech Preview v2, will contain base64-encoded key_id hash
+   ```
+
+4. stateController generates EncryptionConfiguration with hash embedded in socket path:
+   ```yaml
+   apiVersion: apiserver.config.k8s.io/v1
+   kind: EncryptionConfiguration
+   resources:
+     - resources: [configmap]
+       providers:
+         - kms:
+             name: kms-a1b2c3d4e5f67890-configmap-1
+             endpoint: unix:///var/run/kmsplugin/kms-a1b2c3d4e5f67890.socket
+             apiVersion: v2
+   ```
+   The deterministic socket path allows KMS plugin lifecycle management to use the same path.
+
+5. migrationController detects the new secret and initiates re-encryption (no code changes - works with any mode).
+
+6. Resources are re-encrypted using KEK in external KMS via the KMS plugin.
+
+7. conditionController updates status conditions: `EncryptionInProgress`, then `EncryptionCompleted`.
+
+#### Variation: Configuration Changes (Key Rotation)
+
+When cluster admin updates KMS configuration (e.g., new key ARN, different region):
+
+1. keyController recalculates hash from updated APIServer resource.
+2. Compares new hash with hash in most recent encryption key secret annotation.
+3. If hashes differ:
+   - Creates new encryption key secret with new hash
+   - migrationController automatically triggers re-encryption
```

Comment:
We need the File shenanigans to be able to leverage the controllers that manage everything.

How do the KMS provider behave wrt that?
Do they migrate all resources on key rotation? (Do we throttle the requests or is it a DOS attack on KAS?)
Do they query the old KEK?